### PR TITLE
Fix newline matching in URL and email regexes

### LINKED ISSUE #1522: Incorrect Email Validation
https://github.com/marshmallow-code/marshmallow/blob/fbe22eb47db5df64b2c4133f9a5cb6c6920e8dd2/src/marshmallow/validate.py#L136-L151

The email validation regex will match `email@domain.com\n`, `email\n@domain.com`, and `email\n@domain.com\n`.

The issue is that `$` is used to match until the end of a string. Instead, `\Z` should be used. - https://stackoverflow.com/a/48730645

It is possible that other validators might suffer from the same bug, so it would be good if other regexes were also checked.

It is unclear, but this may lead to a security vulnerability in some projects that use marshmallow (depending on how the validator is used), so a quick fix here might be helpful. In my quick look around I didn't notice anything critical, however, so I figured it would be fine to open this issue.

### PR DESCRIPTION
**Pull request by nbanmp**
Resolves #1522

TBH, I didn't test this at all, so we'll see what the test cases say.

### CODE REVIEW

File: src/marshmallow/validate.py [hunk 1]
```diff
@@ -70,7 +70,7 @@ def _regex_generator(self, relative: bool, require_tld: bool):
                         r")?"
                         if relative
                         else r"",  # host is optional, allow for relative URLs
-                        r"(?:/?|[/?]\S+)$",
+                        r"(?:/?|[/?]\S+)\Z",
```
Comment (lafrech) [inline src/marshmallow/validate.py:73]: This line is also on 2.x branch, so I guess we ought to fix it there as well.

https://github.com/marshmallow-code/marshmallow/blob/2.x-line/src/marshmallow/validate.py

AFAICT, MA2's email validation is not impacted.

Comment (deckar01) [reply src/marshmallow/validate.py:73]: I was able to reproduce the same issue on 2.x.

```py
>>> from marshmallow import Schema, fields
>>> 
>>> class Test(Schema):
...     foo = fields.Email()
... 
>>> Test().load({'foo': 'foo@foo.foo\n'})
UnmarshalResult(data={'foo': u'foo@foo.foo\n'}, errors={})
```

Comment (nbanmp) [reply src/marshmallow/validate.py:73]: It looks like `$` is still used for emails on the 2.x branch. https://github.com/marshmallow-code/marshmallow/blob/2.x-line/src/marshmallow/validate.py#L126

Comment (nbanmp) [reply src/marshmallow/validate.py:73]: I can create another PR for 2.x if ya'll want that?

Comment (lafrech) [reply src/marshmallow/validate.py:73]: Good catch. I missed it.

> I can create another PR for 2.x if ya'll want that?

If you don't mind the extra work, that would be nice. The usual process is fix 2.x (if relevant) then merge into dev. Meanwhile, please keep this PR/branch as a reference for the merge into dev.

Thanks.

**Comment by lafrech** [comment]
Please also add a non-regression test.

And add yourself to AUTHORS file.

**Comment by lafrech** [comment]
Thanks. I merged #1524 and merged into `dev` branch, so this is fixed in `dev`.

**Pull request closed without merging**